### ISSUE TITLE
touch support throwing an error in chrome

### ISSUE THREAD

**Issue by falquaddoomi** [processing-js/processing-js#191]
I'm attempting to run a processing.js sketch on a mobile device, but every time the user initiates a mouse event the following exception is thrown:

    Uncaught ReferenceError: eventHandlers is not defined

The error occurs at the following location in processing.js:
https://github.com/processing-js/processing-js/blob/master/processing.js#L7574

I would have thought that 'eventHandlers' would be included in the scope where the callback to attachEventHandlers() is being defined, but apparently not. I've been able to replicate this issue using the mobile device emulator in Chrome for Windows v43.0.2357.124 m.

For completeness, I've included a basic sketch that demonstrates the problem:

	void setup() {
	  size(300, 300); 
	}

	PVector m = new PVector(), displace = new PVector();
	boolean pressing = false;
	void mousePressed() {
	  m.set(mouseX, mouseY);
	  pressing = true;
	}

	void mouseReleased() {
	  pressing = false;
	}

	void draw() {
	  background(0);

	  if (pressing) {
		displace = PVector.sub(m, new PVector(mouseX, mouseY));
		float rad = displace.mag()*2.0;
		fill(255);
		ellipse(m.x, m.y, rad, rad);
	  }
	}

**Comment by Pomax** [comment]
yeah it looks like for touch we're going to have to clean up the code a little and then explain how to use hammer.js or the like on top of Pjs for touch.

**Comment by falquaddoomi** [comment]
Well, I went ahead and hacked at the library a little bit until touch events were working on my mobile device (seems to work on both Chrome for Android and Safari for embedded iOS devices.) The patch is below:

	--- processing.js	Fri Jun 12 13:40:42 2015
	+++ processing_PATCHED.js	Wed Jun 17 11:38:29 2015
	@@ -7476,7 +7476,10 @@
	 /**
	  * Touch and Mouse event handling
	  */
	-module.exports = function withTouch(p, curElement, attachEventHandler, document, PConstants, undef) {
	+module.exports = function withTouch(p, curElement, attachEventHandler, detachEventHandler, eventHandlers, document, PConstants, undef) {
	+
	+  // falquaddoomi: added detachEventHandler, eventHandlers as arguments b/c
	+  // they're used below and were apparently not in scope

	   /**
		* Determine the location of the (mouse) pointer.
	@@ -7658,7 +7661,8 @@
		 }

		 // Refire the touch start event we consumed in this function
	-    curElement.dispatchEvent(t);
	+	// falquaddoomi: disabled to prevent chrome exception re: rethrowing the event
	+    // curElement.dispatchEvent(t);
	   });

	   /**
	@@ -9711,7 +9715,9 @@
		 extend.withCommonFunctions(p);
		 extend.withMath(p);
		 extend.withProxyFunctions(p, removeFirstArgument);
	-    extend.withTouch(p, curElement, attachEventHandler, document, PConstants);
	+    // falquaddoomi: now passing detachEventHandler and eventHandlers
	+    // b/c they're needed within withTouch, but weren't in scope(?)
	+    extend.withTouch(p, curElement, attachEventHandler, detachEventHandler, eventHandlers, document, PConstants);

		 // custom functions and properties are added here
		 if(aFunctions) {

Basically I'm just manually passing in variables to withTouch() that were referenced in the touch handler callback, but for some reason weren't in withTouch()'s scope. I also ended up commenting out the line at the end of the callback where the touch event is re-raised, as it produces the following error in Chrome:

    Uncaught InvalidStateError: Failed to execute 'dispatchEvent' on 'EventTarget': The event is already being dispatched.

I don't know the specifics of how the library works, so I'm not sure if those changes are reasonable, but they seemed to work for me.

**Comment by Pomax** [comment]
can you file this patch as a normal PR? then it can go through the normal review process

**Comment by falquaddoomi** [comment]
Ah, it seems someone else has already submitted a PR with tests as well, something I don't have time to do. Their solution probably works better than mine.

https://github.com/processing-js/processing-js/pull/118

**Comment by Pomax** [comment]
that PR was actually missing tests, but I had forgotten it was there entirely. thanks for reminding

**Comment by menkari** [comment]
Did the original PR ended up getting merged in? working with v1.4.8 and not getting the mouseReleased() and touchEnd() events firing.

**Comment by Pomax** [comment]
1.4.8 is fairly old at this point, can you check against the latest version? (1.4.16, installation using npm is easiest at this point: https://www.npmjs.com/package/processing-js)

**Comment by offbynull** [comment]
I checked out master a few hours ago. This same issue is still occurring.

The patch submitted by @falquaddoomi seems to still work around the issue.
